cli/admin/list.go

```diff
@@ -2,78 +2,49 @@ package admin
 
 import (
 	"fmt"
-	"net"
-	"sort"
+	"time"
 
 	"github.com/apex/log"
-	"github.com/tarantool/cartridge-cli/cli/common"
-	"github.com/tarantool/cartridge-cli/cli/templates"
+	"github.com/tarantool/cartridge-cli/cli/connector"
 )
 
-func adminFuncList(conn net.Conn) error {
-	listResRawMap, err := getFuncListRawMap(conn)
+func adminFuncList(conn *connector.Conn) error {
+	funcInfos, err := getListFuncInfos(conn)
 	if err != nil {
 		return fmt.Errorf("Failed to get functions list: %s", err)
 	}
 
-	if len(listResRawMap) == 0 {
-		return fmt.Errorf("No available admin functions found")
-	}
-
-	funcUsages := make(NameUsages, len(listResRawMap))
-
-	i := 0
-	for funcNameRaw, funcSpecRaw := range listResRawMap {
-		funcName, ok := funcNameRaw.(string)
-		if !ok {
-			return getCliExtError("Functions map key isn't a string: %#v", funcNameRaw)
-		}
-
-		adminFuncSpec, ok := funcSpecRaw.(map[interface{}]interface{})
-		if !ok {
-			return getCliExtError("Function %q spec isn't a map: %#v", funcName, funcSpecRaw)
-		}
-
-		funcUsage, err := getStrValueFromRawMap(adminFuncSpec, "usage")
-		if err != nil {
-			return getCliExtError("Failed to get function %q usage: %s", funcName, err)
-		}
-
-		funcUsages[i] = NameUsage{
-			Name:  funcName,
-			Usage: funcUsage,
-		}
-
-		i++
-	}
-
-	sort.Sort(funcUsages)
-	log.Infof("Available admin functions:\n\n%s", funcUsages.Format())
+	log.Infof("Available admin functions:\n\n%s", funcInfos.FormatUsages())
 
 	return nil
 }
 
-func getFuncListRawMap(conn net.Conn) (map[interface{}]interface{}, error) {
-	adminListFuncBody, err := templates.GetTemplatedStr(&adminListFuncBodyTmpl, map[string]string{
-		"AdminListFuncName": adminListFuncName,
-	})
-
-	listResRaw, err := common.EvalTarantoolConn(conn, adminListFuncBody, common.ConnOpts{})
+func getListFuncInfos(conn *connector.Conn) (*FuncInfos, error) {
+	funcBody, err := getAdminFuncEvalTypedBody(adminListFuncName)
 	if err != nil {
-		return nil, fmt.Errorf("Failed to call %s(): %s", adminListFuncName, err)
+		return nil, err
 	}
 
-	listResRawMap, err := convertToMap(listResRaw)
-	if err != nil {
-		return nil, getCliExtError("Failed to convert %q return value to map", adminListFuncName)
+	req := connector.EvalReq(funcBody).SetReadTimeout(3 * time.Second)
+
+	funcInfosSlice := []FuncInfos{}
+	if err := conn.ExecTyped(req, &funcInfosSlice); err != nil {
+		return nil, err
 	}
 
-	return listResRawMap, nil
+	if len(funcInfosSlice) != 1 {
+		return nil, fmt.Errorf("Function signature received in a bad format")
+	}
+
+	funcInfos := funcInfosSlice[0]
+
+	return &funcInfos, nil
 }
 
 var (
 	adminListFuncBodyTmpl = `
-	local func_help, err = {{ .AdminListFuncName }}('{{ .FuncName }}')
-	return func_help, err
+local func_list, err = {{ .AdminListFuncName }}(...)
+assert(err == nil, err)
+return func_help
```

Comment:
Didn't we add a static filesystem to avoid this (same above)?

Reply:
Yes, now all such Lua code strings can be moved into separated files, but I sure we can perform it in a separated patch.
Added issue https://github.com/tarantool/cartridge-cli/issues/451